hs/src/Reach/Optimize.hs

```diff
@@ -612,6 +631,7 @@ optLet at x e = do
                 meh
   case (extract x, (isPure e && canDupe e), e) of
     (Just dv, True, _) -> doit dv
+    (Just dv, _, DLE_PrimOp {}) -> doit dv
```

Comment:
It looks to me that this is the most important part and you should make a comment about why we are ignoring purity here